### PR TITLE
Grouped conv

### PR DESCRIPTION
**Pull request by duli2012**
A new implementation for unpacked grouped conv kernel, which is around 3x faster than the existing one.

### CODE REVIEW

File: test/data/ops/conv.jsonc [hunk 1]
```diff
@@ -104,10 +104,10 @@
     ]
   },
   {
-    "name": "conv - group - B",
+    "name": "conv - group - C",
```
Comment (xzhu1900) [inline test/data/ops/conv.jsonc:107]: nit: Did we miss a B here?

File: lib/backends/webgl/ops/conv.ts [hunk 2]
```diff
@@ -1,16 +1,153 @@
...
+    const inputSpatialShapeWithPad = inputSpatialShape.map((v, i) => v + adjustPads[i] + adjustPads[i + spatialRank]);
+    const outputSpatialShape =
+        inputSpatialShapeWithPad.map((v, i) => Math.floor((v - dilatedKernelShape[i] + strides[i]) / strides[i]));
+    const outputShape = [batchSize, outChannels].concat(...outputSpatialShape);
+    return outputShape;
+  }
+}
+export class WebGLUnpackedGroupedConv extends Conv implements WebGLOperator {
+  run(inferenceHandler: WebGLInferenceHandler, inputs: Tensor[]): Tensor[] {
+    console.log('here we are.');
```
Comment (xzhu1900) [inline lib/backends/webgl/ops/conv.ts:71]: remove.

File: lib/backends/webgl/ops/conv.ts [hunk 3]
```diff
@@ -1,16 +1,153 @@
...
+    const outputChannelsPerGroup = wShape[0] / this.group;
+    // if kernelShape is not specified in the attributes of this op, infer it from the weight tensor dims
+    if (this.kernelShape.length === 0) {
+      for (let i = 2; i < wShape.length; ++i) {
+        this.kernelShape.push(wShape[i]);
+      }
+    }
+    PoolConvUtil.adjustPadsBasedOnAutoPad(
+        inputs[0].dims, this.strides, this.dilations, this.kernelShape, this.pads, this.autoPad);
+    Logger.verbose(
```
Comment (xzhu1900) [inline lib/backends/webgl/ops/conv.ts:89]: remove?

File: lib/backends/webgl/ops/conv.ts [hunk 4]
```diff
@@ -1,16 +1,153 @@
...
+    const inputTDs = inputs.map((t, i) => handler.getOrCreateTextureData(t, programInfo.inputLayouts[i]));
+    return {
+      inputTextureDatas: inputTDs,
+      outputTextureData: handler.createTextureDataFromLayout(programInfo.outputLayout, inputTDs[0].tensor.type),
+      uniformData: {}
+    };
+  }
+}
+
+export class WebGLUnpackedConv extends Conv {
```
Comment (xzhu1900) [inline lib/backends/webgl/ops/conv.ts:150]: why `WebGLUnpackedConv` doesn't implement `WebGLOperator` as `WebGLUnpackedGroupedConv` do?

Comment (duli2012) [reply lib/backends/webgl/ops/conv.ts:150]: yes, WebGLUnpackedConv consists of two kernels (im2col + matmul), which is not supported by WebGLOperator interface.

Comment (xzhu1900) [reply lib/backends/webgl/ops/conv.ts:150]: ok. make sense.

**Review by xzhu1900 — APPROVED** [review]

File: lib/backends/webgl/ops/conv.ts (hunk 2 above)
Comment (duli2012) [reply lib/backends/webgl/ops/conv.ts:71]: done

File: test/data/ops/conv.jsonc (hunk 1 above)
Comment (duli2012) [reply test/data/ops/conv.jsonc:107]: good catch. ;-)

File: lib/backends/webgl/ops/conv.ts (hunk 3 above)
Comment (duli2012) [reply lib/backends/webgl/ops/conv.ts:89]: are you referring to the log?

**Pull request merged**